F3-229: manifest provider, ec decoupling from f3, and dynamic manifest update

adlrocha opened the pull request:
Partially tackles #229 

## Background
In #337 I did all of the exploration for the implementation end-to-end of #229, but as the PR was growing too big, reviewing would have been quite cumbersome, and rebasing was becoming a bit of a pain, I decided to start dividing the work into smaller self-contained PRs that can (hopefully) be reviewed and merged faster. 

(This will really make my life easier every time that I need to rebase :) ) 

## Implementation
This PR: 
- Decoupled `EcBackend`  and ` FakeEC`  from f3 into its own independent package so it can be used for the integration tests that I was writing in #337 without circular dependencies.
- Implement a `ManifestProvider` interface that abstracts the specific manifest implementation used under the hood. This will hopefully simplify the removal of all the passive testing-specific logic in the future. 
  - There are two implementation of `ManifestProvider`: `Static` , which doesn't accept any runtime changes and is just a bunch of accessors to the underlying manifest, and `DynamicManifest`  which will be the one used to trigger configuration changes in passive testing. 
  - To simplify the review, I haven't pushed the whole implementation of `DynamicManifest`, this will come in the next PR. 
- As suggested by Alex [here](https://github.com/filecoin-project/go-f3/pull/337#discussion_r1635511935), moved network-specific types out of the `gpbft`  protocol 

## Coming Next
- [x] Implementation of `DynamicManifest`  and integration in f3 to trigger configuration rebootstraps.
- [x] Implementation of `ManifestSender`  responsible for periodically broadcasting the current manifest used by the network.
- [x]  Integration tests for f3
- [x]  Implementation of configuration changes without rebootstrap.

masih commented on ec/ec.go:
```diff
@@ -0,0 +1,26 @@
+	"github.com/filecoin-project/go-f3/gpbft"
+)
+
+type ECBackend interface {
```
Avoid repeating package name in type names.

masih commented on ec/ec.go:
```diff
@@ -0,0 +1,26 @@
+	GetPowerTable(context.Context, gpbft.TipSetKey) (gpbft.PowerEntries, error)
+}
+
+type TipSet interface {
```
There is a [`TipSet` type](https://github.com/filecoin-project/go-f3/blob/650861a1a3c36f9103b9010a42ce8e6b0005cb69/gpbft/chain.go#L54) already in `gpbft` package. It would be nice to unify them if they intend to covey the same entity, or rename this one to express that they are indeed different. 

This can be done in a separate PR if it would touch too many lines.

adlrocha replied:
chiming @Kubuxu in here for visibility and to give additional context. This type comes from his EC integration, and I just moved it to its own independent package.

Stebalien replied:
These are, unfortunately, _different_ tipsets.

1. This is an abstraction over the client's concept of a tipset and doesn't, e.g., carry a commitment to a power table etc.
2. The one in the `gpbft` package contains additional commitments, etc.

Honestly, we may just want to rename the one in the `gpbft` package to... ECBlock/ECBlockKey? Or something like that? But that might be even more confusing.

masih commented on f3.go:
```diff
@@ -88,10 +89,10 @@ func (mc *client) Logger() Logger {
-	ds = namespace.Wrap(ds, manifest.NetworkName.DatastorePrefix())
-	cs, err := certstore.OpenOrCreateStore(ctx, ds, 0, manifest.InitialPowerTable)
+func New(ctx context.Context, id gpbft.ActorID, manifest manifest.ManifestProvider, ds datastore.Datastore, h host.Host, manifestServer peer.ID,
+	ps *pubsub.PubSub, sigs gpbft.SignerWithMarshaler, verif gpbft.Verifier, ec ec.ECBackend, log Logger) (*F3, error) {
```
😮‍💨  11 args. I would introduce options here once the implementation settles (but before code freeze).

adlrocha replied:
Agree. Definitely a debt we will want to pay :)

masih commented on f3.go:
```diff
@@ -88,10 +89,10 @@ func (mc *client) Logger() Logger {
-	cs, err := certstore.OpenOrCreateStore(ctx, ds, 0, manifest.InitialPowerTable)
+func New(ctx context.Context, id gpbft.ActorID, manifest manifest.ManifestProvider, ds datastore.Datastore, h host.Host, manifestServer peer.ID,
+	ps *pubsub.PubSub, sigs gpbft.SignerWithMarshaler, verif gpbft.Verifier, ec ec.ECBackend, log Logger) (*F3, error) {
+	ds = namespace.Wrap(ds, manifest.DatastorePrefix())
```
Thinking out loud: should `Manifest` have something as low level as datastore prefix key? We might want to infer it from some other parameter as it seems like lower level impelentation detail as far as `Manifest` goes.

masih commented on gpbft/types_test.go:
```diff
@@ -1,36 +0,0 @@
-			require.True(t, strings.HasSuffix(gotPubSubTopic, string(test.subject)))
-		})
-	}
-}
```
`git mv` to preserve commit history?

adlrocha replied:
Unfortunately, this is not exactly a `git mv` but an aggregation of different tests into the same way. Can history be preserved then? (AFAIK. `git mv` is only for files, right?

masih replied:
I would move the file, then rename it then add the other tests. but no biggy :)

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+	SetManifestChangeCb(OnManifestChange)
+
+	// Manifest accessors
+	MsgPubSubTopic() string
```
I recommend avoiding `Msg` shortened word at this interface api. To use consistent terminology with GPBFT, would it make sense to call this `BroadcastTopic` or simply drop `Msg`?

adlrocha replied:
+1. Dropping `Msg` works. The reason for the prefix is because before we had a different pubsub topic for the manifest (when everything was way less decoupled).

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+	// Returns the list of gpbft options that should be used
+	GpbftOptions() []gpbft.Option
+	// suscribe to manifest updates
+	Subscribe() <-chan struct{}
```
* Document what the expected semantics is: e.g. how many subscribers can subscribe here?
* What about unsubscribe? A pattern you can use is to return an unsubscribe function along with the channel.

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+	// suscribe to manifest updates
+	Subscribe() <-chan struct{}
+	// Set callback for manifest changes
+	SetManifestChangeCb(OnManifestChange)
```
Without the comment above this function it would have taken me loner than it should to infer `Cb` is short for "Callback". 
* I recommend a clearer name and documentation of semantics (similar to what I recommended for `Subscribe`)
* At API design level, it seems odd to have a subscription mechanism to receive notifications, _and_ a hook registry to execute once manifest changes. I am curious if both are necessary and whether the API surface here can be reduced.

adlrocha replied:
I think you may be right, it is not clear yet if I'll need the `Subscribe()`  in the end. Before I needed it to notify the runner that there was a new manifest so it can pick up the new configuration. Now that you mention it, I may be able to integrate this as part of the F3 callback. Unfortunately, that logic comes in the next PR, so if you don't mind, let's revisit this one with the whole logic implemented in the next PR. Thanks!

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+	MsgPubSubTopic() string
+	NetworkName() gpbft.NetworkName
+	BootstrapEpoch() int64
+	DatastorePrefix() datastore.Key
```
Ditto re implementation detail. Is this something that needs to be exposed at interface level?

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+	// Manifest accessors
+	MsgPubSubTopic() string
+	NetworkName() gpbft.NetworkName
+	BootstrapEpoch() int64
```
Thinking out loud: Would it make sense to set bootstrap `TipSet`, where the tipset encapsulates all the info needed for the initial bootstrap (e.g. power table, supplemental data, etc.). That should simplify this API and get rid of having one api per bootstrap-info "stuff".

adlrocha replied:
Fixed by exposing the whole Manifest as discussed in the other comment.

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+	MaxLookaheadRounds   uint64
+}
+
+type EcConfig struct {
```
Does this belong to `ec` package?

adlrocha replied:
This is the EC configuration propagated by the manifest. The manifest is used by the fake EC backend. If I include it as part of that package it leads to a circular dependency.

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+	ECDelay          time.Duration
+	CommiteeLookback uint64
+
+	//Temporary
```
Can you expand on this please? As in to be removed after passive testing? If so, i would leave an explicit `// TODO: `

adlrocha replied:
Chiming @Kubuxu in again. This is not temporary from the passive testing point of view, but I guess the `FakeEC`  one which uses strict timing assumptions.

Kubuxu replied:
I think we should try removing them; they should only be used as part of the FakeEC, which is outliving its purpose. If we have to hardcode them or make the standalone node inoperable, I'm fine with that (as long as we have tests that Alfonso is adding).

I'm fine with us breaking the standalone node for cleanup purposes, especially if it is FakeEC as it will need a lightweight FilecoinAPI client in there.

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+	*EcConfig
+}
+
+func LocalnetManifest() Manifest {
```
use the phrase `localDevnet` for consistency with the rest of the filecoin ecosystem?

adlrocha replied:
fixed.

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+
+// Manifest identifies the specific configuration for
+// the F3 instance currently running.
+type Manifest struct {
```
Dumb question: Why both `ManifestProvider` interface with  verbose per-field API, and `Manifest` struct? Could `ManifestProvider` be a simple one liner interface API that returns `Manifest`?

adlrocha replied:
Good catch, thanks! Initially (before the PRs for EC and other dependencies), I was thinking that for testing providers may need different underlying manifests (so the `StaticProvider` would use a different underlying Manifest that is not the same to the one used by the `DynamicProvider`). It is clear now that this won't be the case, so I will go with your suggestion of directly exposing the Manifest. Thanks!

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+	m := Manifest{
+		NetworkName:    gpbft.NetworkName(fmt.Sprintf("localnet-%X", rng)),
+		BootstrapEpoch: 1000,
+		EcConfig: &EcConfig{
```
Extract to some default EC config `var`.

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+			ECPeriod: 30 * time.Second,
+		},
+	}
+	m.ECBoostrapTimestamp = time.Now().Add(-time.Duration(m.BootstrapEpoch) * m.ECPeriod)
```
Can I ask what this timestamp is used for?

Kubuxu replied:
It is used for producing new blocks/new head over time. We should probably abstract out time for tests.

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+	// starting from BootstrapEpoch.
+	PowerUpdate []certs.PowerTableDelta
+	// Config parameters for gpbft
+	*GpbftConfig
```
Why top level field? Ditto for `EcConfig`.

adlrocha replied:
Make the access less verbose, but no strong opinions here.

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+
+type Version string
+
+type GpbftConfig struct {
```
Simply take `gopbft.Option` as a field of `Manifest`?

adlrocha replied:
This may not be possible (unless I am missing something). We need the Manifest to be serializable so it can me persisted and moved around, and using `gpbft.Option` required implementing all the mechanics to make them JSON serializable.

masih commented on manifest/manifest.go:
```diff
@@ -0,0 +1,150 @@
+func (m Manifest) GpbftOptions() []gpbft.Option {
+	var opts []gpbft.Option
+
+	if m.Delta != 0 {
```
These `if` checks for default are technically incorrect? i.e. don't cover a case where user explicitly sets something to zero value. 

I recommend thinking about if that matters here, and if so use pointers for fields instead and check if they are `nil`. Better yet, avoid it all by taking options directly

adlrocha replied:
I updated it a bit allowing zero to be set except for `Delta`  which doesn't make much sense.

masih commented on manifest/manifest_test.go:
```diff
@@ -0,0 +1,114 @@
+
+}
+
+func TestNetworkName(t *testing.T) {
```
`TestManifest_NetworkName` ? This is a convention that so far has worked well for me. Ditto for the rest of the tests. At a very least i would include `Manifest` in the test name, since that's what it is testing.

masih commented on manifest/static.go:
```diff
@@ -0,0 +1,56 @@
+
+// Static manifest provider that doesn't allow any changes
+// in runtime to the initial manifest set in the provider
+type Static struct {
```
`StaticManifestProvider`?

masih commented on manifest/static.go:
```diff
@@ -0,0 +1,56 @@
+
+var _ ManifestProvider = (*Static)(nil)
+
+func NewStaticManifest(m *Manifest) ManifestProvider {
```
Ditto re naming, i.e. include `Provider` suffix.

masih commented on passive/dynamic_manifest.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+func (m *DynamicManifest) Subscribe() <-chan struct{} {
+	return m.manifestUpdates
```
This means only one of the subscribers will receive the update. Since the intended semantics of `Subscribe` is undocumented I don't know if this is intended.

Kubuxu commented on manifest/manifest.go:
```diff
@@ -0,0 +1,161 @@
+	// This is used to identify if a new config needs to be applied
+	Sequence uint64
+	// Initial instance to use for gpbft
+	InitialInstance uint64
```
This is the property of the whole F3, not only gpbft. We would have teardown F3 and certstore to change it.

Kubuxu commented on manifest/manifest.go:
```diff
@@ -0,0 +1,161 @@
+	// Initial power table used when bootstrapping. This is used for test
+	// networks or if we want to force a specific power table when
+	// rebootstrapping. If nil, the power table is fetched from the host
+	InitialPowerTable gpbft.PowerEntries
```
InitialPowerTable does not exist anymore. It is fetched from EC. This is another remnant of FakeEC.

adlrocha replied:
Fixed

Kubuxu commented on manifest/manifest.go:
```diff
@@ -0,0 +1,161 @@
+		return "", xerrors.Errorf("computing manifest version: %w", err)
+	}
+	return Version(hex.EncodeToString(gpbft.MakeCid(b))), nil
+}
```
I don't like relying on stable serialization of JSON. I see that we use it for the topic name. Can we use the sequence number of it instead?

Kubuxu commented on passive/dynamic_manifest.go:
```diff
@@ -0,0 +1,222 @@
+				}
+
+				// if the upgrade epoch is reached or already passed.
+				if ts.Epoch() >= m.nextManifest.BootstrapEpoch {
```
What about the case where we didn't boostrap yet? we might want to swap out the manifest either way

Kubuxu reviewed (commented):
Notes:
- changing the pubsub topic name will run into issues in Lotus because Lotus has a whitelist of pubsub topics which are allowed to be used, not sure why exactly but it will be a problem

masih commented on test/f3_test.go:
```diff
@@ -0,0 +1,545 @@
+func (e *testEnv) updateManifest() {
+	e.manifestSender.UpdateManifest(&e.manifest)
+	// wait for the manifest to propagate
+	time.Sleep(2 * ManifestSenderTimeout)
```
What indicates that the manifest has propagated? Wrap that logic in a `require.Eventually` instead of arbitrary wait time.

masih reviewed (commented):
Blockers:
* flow of `setupGpbftRunner` when errors occur
* arbitrary timeouts in tests.

Left some suggestions. Thank you for powering through this @adlrocha 💪

masih commented on f3.go:
```diff
@@ -254,102 +326,148 @@ func (m *F3) GetCert(ctx context.Context, instance uint64) (*certs.FinalityCerti
-		runnerErrCh <- err
-	}()
+	case err = <-runnerErrCh:
+		return err
```
Right you are.

masih commented on f3.go:
```diff
@@ -183,35 +202,88 @@ func (m *F3) setupPubsub(runner *gpbftRunner) error {
+	m.client.messageQueue = messageQueue
+
+	go func() {
+		latest := m.certStore.Latest()
```
I think not doing it means instance gets started even if context is cancelled. We probably want to check the context eitehr here or earlier in Run.

adlrocha replied:
You are right, fixed.

masih commented on f3.go:
```diff
@@ -183,35 +202,88 @@ func (m *F3) setupPubsub(runner *gpbftRunner) error {
+		return
+	}
+
+	messageQueue := make(chan gpbft.ValidatedMessage, 20)
```
Can you add the above comment to the code please?

adlrocha commented:
@masih, I think I addressed all of your comments, rebased again, and fixed tests. If you could have a final look and don't have any big blocker it would be great if we could get this one merged. It is getting harder and harder to maintain this branch with every new commit as the PR is getting huge :see_no_evil:. Thanks :pray:

masih commented on manifest/dynamic_manifest.go:
```diff
@@ -0,0 +1,226 @@
+			msg, err = manifestSub.Next(ctx)
+			if err != nil {
+				if ctx.Err() != nil {
+					break
```
Should this be `break loop`?

masih commented on manifest/dynamic_manifest.go:
```diff
@@ -0,0 +1,226 @@
+	}
+
+loop:
+	for {
```
It would make the logic simpler to add a condition here to check if contex has error or not: `for ctx.Err() == nil {`

masih commented on manifest/dynamic_manifest.go:
```diff
@@ -0,0 +1,226 @@
+	}
+
+	manifestSub.Cancel()
+	if err := m.teardownManifestPubsub(); err != nil {
```
Cancel and tear down could go in a `differ` function to further simplify the code path if this function will keep doing its thing until context is done.

masih commented on manifest/dynamic_manifest.go:
```diff
@@ -0,0 +1,226 @@
+	if manifest.Sequence <= m.manifest.Sequence {
+		return
+	}
+	// TODO: Any additional logic here to determine what manifests to accept or not?
```
Double checking that the manifest signature/sender is confirmed to be the passive test control server before we hit this.

masih reviewed (approved):
The failing tests with race detector need to be resolved. 
Otherwise Looks good thanks @adlrocha 👍 

Left some suggestions.
